src/Fetch.js

```diff
@@ -438,6 +441,39 @@ function __emscripten_fetch_xhr(fetch, onsuccess, onerror, onprogress) {
     if (xhr.statusText) stringToUTF8(xhr.statusText, fetch + {{{ C_STRUCTS.emscripten_fetch_t.statusText }}}, 64);
     if (onprogress) onprogress(fetch, xhr, e);
   }
+  xhr.onreadystatechange = function(e) {
+      HEAPU16[fetch + {{{ C_STRUCTS.emscripten_fetch_t.readyState }}} >> 1] = xhr.readyState;
+      if (xhr.readyState == 2 && fetchAttrResponseHeaders && HEAPU32[fetch + {{{ C_STRUCTS.emscripten_fetch_t.responseHeaders }}} >> 2] == 0) {
+        HEAPU16[fetch + {{{ C_STRUCTS.emscripten_fetch_t.status }}} >> 1] = xhr.status;
+        var headers = xhr.getAllResponseHeaders();
+        var headerList = headers.trim().split(/[\r\n]+/);
+        var numHeaders = headerList.length;
+        if (numHeaders > 0) {
+          var allocBytes = (numHeaders * 2 + 1) * 4;
+          arrayPtr = _malloc(allocBytes);
+          HEAPU32[fetch + {{{ C_STRUCTS.emscripten_fetch_t.responseHeaders }}} >> 2] = arrayPtr;
+          HEAPU32[arrayPtr + (numHeaders * 8) >> 2] = 0;
+        }
+        for (var i = 0; i < numHeaders; ++i) {
+          var parts = headerList[i].split(": ");
+          var headerKey = parts.shift();
+          var keyLength = lengthBytesUTF8(headerKey)+1;
+          var keyPtr = _malloc(keyLength);
+          stringToUTF8(headerKey, keyPtr, keyLength);
+
+          var headerValue = parts.join(": ");
+          var valueLength = lengthBytesUTF8(headerValue)+1;
+          var valuePtr = _malloc(valueLength);
+          stringToUTF8(headerValue, valuePtr, valueLength);
+
+          var memoryOffset = i * 8;
+          HEAPU32[arrayPtr + memoryOffset >> 2] = keyPtr;
+          HEAPU32[arrayPtr + (memoryOffset + 4) >> 2] = valuePtr;
+        }
+      }
+      xhr.onreadystatechange = null;
+      if (onheadersreceived) onheadersreceived(fetch, xhr, e);
```

Comment:
The contents of `onreadystatechange` here that unpack the response headers adds too much code size. Instead of always unconditionally unpacking the headers for all users here, let's instead add two functions:

```c
size_t emscripten_fetch_get_response_headers_length(emscripten_fetch_t *fetch);
size_t emscripten_fetch_get_response_headers(emscripten_fetch_t *fetch, char *dst, size_t dstSizeBytes);
```
which copies the string `xhr.getAllResponseHeaders()` to the given `dst` pointer, truncated to at most `dstSizeBytes`, null-terminated, and returns the number of bytes that were written.

Then, if we want to offer functionality to unpack, let's add another utility function
```c
char **emscripten_fetch_unpack_response_headers(char *headersString);
void emscripten_fetch_free_unpacked_response_headers(char **unpackedHeaders);
```
which performs the unpacking and invokes `malloc()` etc, and documents that one needs to call `emscripten_fetch_free_unpacked_response_headers()` to free up the returned pointer.

This way users who do not need to read response headers, and/or users who do not need to rely on response headers being unpacked to an array of strings will not get bloated up code size due to features that they don't need.

Reply:
Done.

Reply:
You might want to check how I've done it - I wasn't sure what the best way was of inter-oping with JavaScript. I added some hidden functions to the library_fetch "header" and call those from inside the cpp. I suppose we could go a step further and wrap them in EM_ASM blocks.